### PR TITLE
New Merkle tree implementation

### PR DESCRIPTION
**Pull request by irakliyk**
This PR migrates the codebase to the new `MerkleTree` implementation as described in #8.

- [x] Implement batch Merkle proof serialization for new Merkle tree
- [x] Update random element generator to use new `Hasher` trait
- [x] Migrate the codebase to the new Merkle tree implementation
- [x] Update Merkle proof verification example
- [x] Update comments and docs

### CODE REVIEW

File: crypto/src/hash/mod.rs [hunk 1]
```diff
@@ -35,13 +52,31 @@ impl Hasher for Blake3_256 {
...
         let bytes = E::elements_as_bytes(elements);
         blake3::hash(&bytes).into()
     }
 
-    fn hash_fn() -> HashFunction {
-        blake3
+    fn merge_with_int(seed: Self::Digest, value: u64) -> Self::Digest {
+        let mut data = [0; 64];
+        data[..32].copy_from_slice(&seed);
+        data[56..].copy_from_slice(&value.to_le_bytes());
```
Comment (kchalkias) [inline crypto/src/hash/mod.rs:71]: Ooc, why did we choose to append the u64 value at the end and not use data[40]? Same for sha3.

Comment (irakliyk) [reply crypto/src/hash/mod.rs]: no good reason. fixed!

File: common/src/proof/mod.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        // conjectured security requires half the queries as compared to proven security
+        let num_queries = if optimistic {
+            options.num_queries()
+        } else {
+            options.num_queries() / 2
+        };
+
+        let one_over_rho =
+            (options.blowup_factor() / self.context.ce_blowup_factor as usize) as u32;
+        let security_per_query = 31 - one_over_rho.leading_zeros(); // same as log2(one_over_rho)
```
[the remark this answers is not in the document]
Comment (irakliyk) [reply common/src/proof/mod.rs]: The extra bit was accounted for in the number of queries computation above. But this was super confusing, and proven security estimate wasn't accurate (the conjectured security was fine). So, I refactored proof security computations.

File: crypto/src/hash/mod.rs [hunk 3]
```diff
@@ -68,7 +68,7 @@ impl Hasher for Blake3_256 {
     fn merge_with_int(seed: Self::Digest, value: u64) -> Self::Digest {
         let mut data = [0; 64];
```
Comment (kchalkias) [inline crypto/src/hash/mod.rs:69]: Prefer `let mut data = [0; 40]`, then `data[32..].copy_from_slice` for both hash schemes.

**Review by kchalkias — APPROVED** [review]
LGTM, just a micro-optimization comment

File: crypto/src/hash/mod.rs (hunk 3 above)
Comment (irakliyk) [reply crypto/src/hash/mod.rs]: fixed!

**Pull request merged**